client/lib/use-in-view/index.ts

```diff
@@ -0,0 +1,46 @@
+import { RefObject, useEffect, useRef } from 'react';
+
+export function useInView< T extends Element >( oneTimeCallback: () => void ): RefObject< T > {
+	const viewedRef = useRef< boolean >( false );
+	const elementRef = useRef< T >( null );
+	const observerRef = useRef< IntersectionObserver >();
+
+	useEffect( () => {
+		// We can't do anything without a valid reference to an element on the page
+		if ( ! elementRef.current ) {
+			return;
+		}
+
+		// If the observer is already defined, no need to continue
+		if ( observerRef.current ) {
+			return;
+		}
+
+		const handler = ( entries: IntersectionObserverEntry[] ) => {
+			// Only fire once per page load
+			if ( viewedRef.current ) {
+				return;
+			}
+
+			const [ entry ] = entries;
+			if ( ! entry.isIntersecting ) {
+				return;
+			}
+
+			oneTimeCallback();
+
+			viewedRef.current = true;
```

Comment:
Does it make sense to not connect the intersection observer if the callback has already been fired? I think it'll save some cycles as there will be no observer installed at all.

Reply:
I think this would only be beneficial if the component unmounts, so I didn't make a change. There are already a few optimisations in place.